File: multi_roblox/servers.py
"""Roblox public game/server API helpers for server hopping."""
import re
from typing import Optional

import requests

_SESSION = requests.Session()
_SESSION.headers.update({"User-Agent": "Roblox/WinInet", "Accept": "application/json"})

_PLACE_URL_RE = re.compile(r"/games/(\d+)")


def parse_place_id(value: str) -> Optional[int]:
    """Accept a numeric placeId or a roblox.com /games/<id>/ URL."""
    value = value.strip()
    if value.isdigit():
        return int(value)
    m = _PLACE_URL_RE.search(value)
    return int(m.group(1)) if m else None


def fetch_servers(place_id: int, limit: int = 100, sort: str = "Asc"):
    """Return the public servers list for a place. Raises on network failure."""
    url = f"https://games.roblox.com/v1/games/{place_id}/servers/Public"
    resp = _SESSION.get(url, params={"sortOrder": sort, "limit": limit}, timeout=10)
    resp.raise_for_status()
    return resp.json().get("data", [])


def pick_server(place_id: int, exclude_job_ids=()) -> Optional[dict]:
    """Choose a non-full public server that isn't in the exclude set.

    Prefers servers with the most headroom but at least one player, which gives
    a smoother hop than dropping into a brand-new empty server.
    """
    exclude = set(exclude_job_ids or ())
    candidates = []
    for srv in fetch_servers(place_id):
        job_id = srv.get("id")
        if not job_id or job_id in exclude:
            continue
        playing = srv.get("playing", 0)
        max_players = srv.get("maxPlayers", 0)
        if max_players and playing < max_players:
            candidates.append((max_players - playing, playing, srv))
    if not candidates:
        return None
    # prefer non-empty servers with the most slots free
    candidates.sort(key=lambda c: (c[1] == 0, -c[0]))
    return candidates[0][2]


def join_uri(place_id: int, job_id: str) -> str:
    return f"roblox://placeId={place_id}&gameInstanceId={job_id}"


def launch_uri(place_id: int) -> str:
    return f"roblox://placeId={place_id}"

File: csharp/MultiRobloxManager/Services/Servers.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace MultiRobloxManager.Services;

internal static class Servers
{
    private static readonly Regex PlaceUrlRe = new(@"/games/(\d+)", RegexOptions.Compiled);
    private static readonly HttpClient SharedClient = BuildSharedClient();

    private static HttpClient BuildSharedClient()
    {
        var c = new HttpClient { Timeout = TimeSpan.FromSeconds(10) };
        c.DefaultRequestHeaders.UserAgent.ParseAdd("Roblox/WinInet");
        c.DefaultRequestHeaders.Accept.ParseAdd("application/json");
        return c;
    }

    public static long? ParsePlaceId(string value)
    {
        value = (value ?? "").Trim();
        if (string.IsNullOrEmpty(value)) return null;
        if (long.TryParse(value, out var n)) return n;
        var m = PlaceUrlRe.Match(value);
        return m.Success ? long.Parse(m.Groups[1].Value) : null;
    }

    /// <summary>Build a roblox:// URI for the protocol-handler join.</summary>
    public static string JoinUri(long placeId, string jobId)
        => $"roblox://placeId={placeId}&gameInstanceId={jobId}";

    public static string LaunchUri(long placeId)
        => $"roblox://placeId={placeId}";

    public sealed record ServerInfo(string Id, int Playing, int MaxPlayers);

    public static async Task<List<ServerInfo>> FetchServersAsync(
        long placeId, int limit = 100, string? proxy = null)
    {
        var url = $"https://games.roblox.com/v1/games/{placeId}/servers/Public?sortOrder=Asc&limit={limit}";
        HttpResponseMessage resp;
        if (!string.IsNullOrEmpty(proxy))
        {
            var handler = new HttpClientHandler
            {
                Proxy = new WebProxy(proxy),
                UseProxy = true,
            };
            using var client = new HttpClient(handler) { Timeout = TimeSpan.FromSeconds(10) };
            client.DefaultRequestHeaders.UserAgent.ParseAdd("Roblox/WinInet");
            resp = await client.GetAsync(url);
        }
        else
        {
            resp = await SharedClient.GetAsync(url);
        }
        resp.EnsureSuccessStatusCode();
        var body = await resp.Content.ReadAsStringAsync();
        using var doc = JsonDocument.Parse(body);
        var list = new List<ServerInfo>();
        if (!doc.RootElement.TryGetProperty("data", out var data)) return list;
        foreach (var s in data.EnumerateArray())
        {
            var id = s.TryGetProperty("id", out var idE) ? idE.GetString() : null;
            if (string.IsNullOrEmpty(id)) continue;
            int playing = s.TryGetProperty("playing", out var pE) ? pE.GetInt32() : 0;
            int max = s.TryGetProperty("maxPlayers", out var mE) ? mE.GetInt32() : 0;
            list.Add(new ServerInfo(id!, playing, max));
        }
        return list;
    }

    /// <summary>Pick a non-full public server not in the exclude set.</summary>
    public static async Task<ServerInfo?> PickServerAsync(
        long placeId, IEnumerable<string>? excludeJobIds = null, string? proxy = null)
    {
        var exclude = new HashSet<string>(excludeJobIds ?? Array.Empty<string>());
        var all = await FetchServersAsync(placeId, proxy: proxy);
        var candidates = all
            .Where(s => !exclude.Contains(s.Id) && s.MaxPlayers > 0 && s.Playing < s.MaxPlayers)
            .Select(s => new { Free = s.MaxPlayers - s.Playing, s.Playing, Server = s })
            .OrderBy(x => x.Playing == 0) // prefer non-empty
            .ThenByDescending(x => x.Free)
            .ToList();
        return candidates.Count > 0 ? candidates[0].Server : null;
    }
}
